0011-R-multiple_reminders_actors_impl.md

```diff
@@ -0,0 +1,149 @@
+# Allow Multiple Versions of Reminder/Actors Implementation
+
+* Author(s): Deepanshu Agarwal (@DeepanshuA), Alessandro Segala (@ItalyPaleAle)
+* State: Ready for Implementation
+* Updated: 2023-12-28
+
+## Overview
+
+As a user, I should be able to use Dapr core functionality of Actors/reminders i.e. leverage Actors/reminders interfaces BUT to be able to implement these interfaces in a new Control Plane service, which as a user, I should be allowed to design as per the way it solves issues for me.
+This is a proposal to allow multiple versions of Reminders/Actors Implementation.
+
+## Background
+
+Currently, we see multiple issue with Reminders creation/deletion at scale and with placement Service.
+We should be able to allow users to use core reminders/actors functionality of Dapr runtime; while implementing Reminders/Actors interface in their own way to solve these issues.
+
+This would let users solve reminders/actors issues as per their own requirements and design.
+This would also let users design their solution in an independent way.
+
+## Related Items
+
+This has already been discussed in different maintainers meetings, notably in [this issue](https://github.com/dapr/dapr/issues/7203).
+
+## Implementation Details
+
+### Design
+
+Reminders / Actors should have their Interfaces segregated in such a way that it shuld be possible to implement them in an independent way.
+There are two places where APIs or functions required for Actors and Reminders are defined.
+
+1. Protos in dapr/proto package.
+
+2. Interfaces in pkg/actors/internal for placement and reminder.
+
+Out of these 2, protos are not to be updated, as every implementation will have their own RPCs defined accordingly.
+
+Only interfaces in pkg/actors/internal have to be updated. A major work has already been done for them in release 1.12 already by segregating actors, reminders and timers interfaces (https://github.com/dapr/dapr/pull/6669).
+
+```go
+
+// PlacementService allows for interacting with the actor placement service.
+type PlacementService interface {
+	io.Closer
+
+	Start(context.Context) error
```

Comment:
This API seems to be biased towards an alternative implementation. Is there a way to show how the existing placement service will work with this interface? Is there a way to elevate the abstraction to a small set of methods:
- Start()
- WaitUntilReady()
- Lookup()
- RegisterActorType()

Or clarify what each of the other methods are needed for. If the interface is simply an union of the methods of each possible implementation, then the interface is not abstracting them properly and each vendor will need to add to this list of methods.

Reply:
These interfaces are taken as it is directly from current dapr/dapr master.
Existing placement client works with this interface.
There are two methods which are not used right now: ReportActorDeactivation and HaltActorFn, but they are useful when actors lifecycle information needs to be updated in Placement service, if it maintains them.

I have added comments with these APIs.